packages/iframe-stamper/src/index.ts

```diff
@@ -0,0 +1,182 @@
+      throw new Error("Cannot initialize iframe in non-browser environment");
+    }
+
+    if (document.getElementById(config.iframeElementId)) {
+      throw new Error(
+        `Iframe element with ID ${config.iframeElementId} already exists`
+      );
+    }
+
+    const container = document.getElementById(config.iframeContainerId);
+    if (!container) {
+      throw new Error(
+        `Cannot create iframe stamper: no container with ID ${config.iframeContainerId} exists in the current document`
+      );
+    }
+    this.container = container;
+
+    var iframe = window.document.createElement("iframe");
+    iframe.id = config.iframeElementId;
+    iframe.src = config.iframeUrl;
+    this.iframe = iframe;
+
+    const iframeUrl = new URL(config.iframeUrl);
+    this.iframeOrigin = iframeUrl.origin;
+
+    // This is populated once the iframe is ready. Call `.init()` to kick off DOM insertion!
+    this.iframePublicKey = null;
+  }
+
+  /**
+   * Inserts the iframe on the page and returns a promise resolving to the iframe's public key
+   */
+  async init(): Promise<string> {
+    this.container.appendChild(this.iframe);
+    return new Promise((resolve, _reject) => {
+      window.addEventListener(
+        "message",
+        (event) => {
+          if (event.origin !== this.iframeOrigin) {
+            // There might be other things going on in the window, for example: react dev tools, other extensions, etc.
+            // Instead of erroring out
+            return;
+          }
+          if (event.data && event.data["type"] == "PUBLIC_KEY_READY") {
+            resolve(event.data["value"]);
+            this.iframePublicKey = event.data["value"];
+          }
+        },
+        false
+      );
+    });
+  }
+
+  /**
+   * Removes the iframe from the DOM
+   */
+  clear() {
+    this.iframe.remove();
+  }
+
+  /**
+   * Returns the public key, or `null` if the underlying iframe isn't properly initialized.
+   */
+  publicKey(): string | null {
+    return this.iframePublicKey;
+  }
+
+  /**
+   * Function to inject a new credential into the iframe
+   * The bundle should be encrypted to the iframe's initial public key
+   * Encryption should be performed with HPKE (RFC 9180).
+   * This is used during recovery flows.
+   */
+  async injectRecoveryBundle(bundle: string): Promise<boolean> {
+    this.iframe.contentWindow?.postMessage(
+      {
+        type: "INJECT_RECOVERY_BUNDLE",
+        value: bundle,
+      },
+      "*"
+    );
+
+    return new Promise((resolve, _reject) => {
+      window.addEventListener(
+        "message",
+        (event) => {
+          if (event.origin !== this.iframeOrigin) {
+            // There might be other things going on in the window, for example: react dev tools, other extensions, etc.
+            // Instead of erroring out we simply return. Not our event!
+            return;
+          }
+          if (event.data && event.data["type"] == "BUNDLE_INJECTED") {
```

Comment:
tiny nit - same optional chaining pattern here (don't feel strongly).

could be useful to throw some of these types into consts in a single place